app/models/friend_request.rb

```diff
@@ -0,0 +1,31 @@
+class FriendRequest < ApplicationRecord
+  belongs_to :sender, class_name: 'User'
+  belongs_to :receiver, class_name: 'User'
+ 
+  # validate :not_friends
+  validate :not_pending
+  validate :not_self
+  validate :not_friends
+
+  def accept
+    sender.basic_friends << receiver
+    destroy
+  end
+
+  private
+
+  def not_self
+    errors.add(:receiver, "can't be equal to user") if sender == receiver
+  end
+
+  def not_pending
```

Comment:
I would rather have this check at the database level also. What if for instance there are two users in different parts of the world and they're in each other's profile? Now both of them wanna add the other and they both clicked the "Add" button at exactly the same time. These may cause failure in validating the consecutive requests. You area already making one side of this record unique in your database. You just need to make sure that the other side also happens (i.e, if user1 sends a request to user2 then at the database level we should also ensure that user2 cannot send a request to user1). 

At this moment you are carrying out the checking here at the model level. This isn't bad in itself but probably not enough to cover for the scenario that I mentioned above.

Reply:
Okay, but rails will not throw an error when the id's are opposite, how do we check this on the database level?

Reply:
Excellent question @YellowPipe (and @zinagithub)!

Ok, I know you guys are awesome engineers. So all I will do is give you some clue. If you observe, in your `schema.rb` you have the following line in your `friend_requests` table:

```ruby
t.index ["receiver_id", "sender_id"], name: "index_friend_requests_on_receiver_id_and_sender_id", unique: true
```
Now, that means if we already have the ff record in the `friend_requests` table:
![image](https://user-images.githubusercontent.com/6015315/59075248-38f48680-8902-11e9-9e4e-51c87b40bac7.png)

This means when we do the ff:
```
ryto = User.find(1)
jim = User.find(3)

ryto.friend_requests_sent.create(friend_id: jim.id)
```
We will get the ff:

![image](https://user-images.githubusercontent.com/6015315/59075352-ae605700-8902-11e9-9d15-d61794a89eef.png)

The question is how to make sure the ff also gives us the same error?:

```ruby
jim.friend_requests_sent.create(receiver_id: ryto.id)
```
Ok two clues:

1. We can create a private method that will check whether there's already a friend request existing between `ryto` and `jim`. If yes, then it should add an error. You can research on how to add errors in model validation.

2. We can then have a `before_validation` callback (or `before_save`, take your pick) that will invoke this private method. This way when one of either `ryto` or `jim` sends a request to each other (when there's already an existing request between them), we will get the same error as in the picture.

**PS: You can and should also do the same for Friendship. When there's an existing friendship between two users, neither of them should be able to create a duplicate friendship with the same friend.**